internal/wantrequestmanager/wantrequestmanager.go

```diff
@@ -0,0 +1,453 @@
+		return wanted
+	}
+
+	// Inform the BlockPresenceManager of any HAVEs / DONT_HAVEs
+	wrm.bpm.ReceiveFrom(msg.From, msg.Haves, msg.DontHaves)
+
+	// Inform interested WantRequests
+	for wr := range interested {
+		wrWanted := wr.receiveMessage(msg)
+
+		// Record which WantRequests wanted the blocks in the message
+		_ = wrWanted.ForEach(func(c cid.Cid) error {
+			wanted.Add(c)
+			return nil
+		})
+	}
+	return wanted
+}
+
+// Called when the WantRequest has been closed
+func (wrm *WantRequestManager) removeWantRequest(wr *WantRequest) {
+	wrm.lk.Lock()
+	defer wrm.lk.Unlock()
+
+	// Remove WantRequest from WantRequestManager
+	for c := range wr.ks {
+		kwrs := wrm.wrs[c]
+
+		delete(kwrs, wr)
+		if len(kwrs) == 0 {
+			delete(wrm.wrs, c)
+		}
+	}
+}
+
+// Keys are in the wanted state until the block is received and they change to
+// the unwanted state.
+type ReqKeyState bool
+
+const (
+	ReqKeyWanted   ReqKeyState = true
+	ReqKeyUnwanted ReqKeyState = false
+)
+
+// A WantRequest keeps track of all the wants for a Session.GetBlocks(keys)
+// call.
+type WantRequest struct {
+	// channel of received blocks
+	Out chan blocks.Block
+	// reference to the WantRequestManager
+	wrm *WantRequestManager
+	// messages is a channel of incoming messages received by Bitswap
+	messages chan *messageWanted
+	// closedCh is used to signal when the WantRequest has closed, so as to
+	// stop sending on the messages channel
+	closedCh chan struct{}
+
+	// Locks the variables below
+	lk sync.RWMutex
+	// the keys this WantRequest is interested in and their state:
+	// wanted / unwanted
+	ks map[cid.Cid]ReqKeyState
+	// closed indicates whether the WantRequest has been cancelled
+	closed bool
+}
+
+// Creates a new WantRequest for the given keys.
+func newWantRequest(wrm *WantRequestManager, ks []cid.Cid) *WantRequest {
+	wr := &WantRequest{
+		wrm:      wrm,
+		ks:       make(map[cid.Cid]ReqKeyState, len(ks)),
+		messages: make(chan *messageWanted, len(ks)),
+		Out:      make(chan blocks.Block, len(ks)),
+		closedCh: make(chan struct{}),
+	}
+
+	for _, c := range ks {
+		// Initialize the want state to "wanted"
+		wr.ks[c] = ReqKeyWanted
+	}
+
+	return wr
+}
+
+// Called when an incoming message arrives that has blocks / HAVEs / DONT_HAVEs
+// for the keys in this WantRequest
+func (wr *WantRequest) receiveMessage(msg *IncomingMessage) *cid.Set {
+	wr.lk.Lock()
+
+	// If the session or want request has been cancelled, bail out
+	if wr.closed {
+		wr.lk.Unlock()
+		return cid.NewSet()
+	}
+
+	// Filter incoming message for blocks / HAVEs / DONT_HAVEs that this
+	// particular WantRequest is interested in
+	blks, wanted := wr.receiveBlocks(msg.Blks)
+	fmsg := &IncomingMessage{
+		From:      msg.From,
+		Blks:      blks,
+		Haves:     wr.filterKeys(msg.Haves),
+		DontHaves: wr.filterKeys(msg.DontHaves),
+	}
+
+	wr.lk.Unlock()
+
+	// Send the message and the set of cids of wanted blocks
+	select {
+	case wr.messages <- &messageWanted{fmsg, wanted}:
```

Comment:
If there's _any_ way to avoid this event loop, I'd do that. We're mixing locked state with event loops.

Reply:
Looking at this, I'm pretty sure we don't need this event loop at all.

Reply:
You're right it would be better not to mix locking with event loops.

We need locking so that we can return which blocks were wanted from `PublishToSessions()`

I believe we need an event loop to
- send blocks on the outgoing channel (returned from Session.GetBlocks())
- listen for `requestCtx.Done()` and `sessCtx.Done()`

In future we'll probably also need it to read from the incoming channel of keys.

Any way we can avoid that?

Reply:
The outgoing channel is completely buffered, right? Can't we just write into it directly?

Reply:
Ah, ok, if we're streaming wants, this will no longer apply.